### PR TITLE
feat: ability tree navigation using scroll wheel

### PR DESCRIPTION
**Pull request by byBackfish**
Adds the ability to use the scroll wheel to navigate up and down in the ability tree

Suggestion: [add scrolling support to ability tree](https://feedback.wynntils.com/posts/102/add-scrolling-support-to-ability-tree)

### CODE REVIEW

File: src/main/java/com/wynntils/modules/utilities/events/ClientEvents.java [hunk 1]
```diff
@@ -1347,4 +1349,29 @@ public void onHeldItemChange(PacketEvent<CPacketHeldItemChange> e) {
...
+    @SubscribeEvent
+    public void onAbilityTreeScroll(GuiOverlapEvent.ChestOverlap.HandleMouseInput e) {
+        if (!Reference.onWorld) return;
+        if (!UtilitiesConfig.INSTANCE.shouldAbilityScroll) return;
+
+        int scrollAmount = Mouse.getDWheel() / 120;
+        if (scrollAmount == 0) return;
+
+        IInventory lowerInv = e.getGui().getLowerInv();
+        if (!lowerInv.getName().endsWith("Abilities")) return;
```
Comment (Incompleteusern) [inline src/main/java/com/wynntils/modules/utilities/events/ClientEvents.java]: Can this be faked with bank names?

Comment (byBackfish) [reply src/main/java/com/wynntils/modules/utilities/events/ClientEvents.java]: i don't think so, but I added a regex pattern check now 👍

File: src/main/java/com/wynntils/modules/utilities/configs/UtilitiesConfig.java [hunk 2]
```diff
@@ -79,6 +79,12 @@ public class UtilitiesConfig extends SettingsClass {
     @Setting(displayName = "Clicking on Locked Items", description = "Should moving items to and from locked inventory slots be prevented?", order = 5)
     public boolean preventSlotClicking = false;
 
+    @Setting(displayName = "Scrolling in Ability Tree", description = "Should using the scroll wheel in the ability tree move a page down/up?", order = 5)
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/configs/UtilitiesConfig.java]: Order is same as the config before

File: src/main/java/com/wynntils/modules/utilities/events/ClientEvents.java [hunk 3]
```diff
@@ -111,6 +113,7 @@ public class ClientEvents implements Listener {
     private static final Pattern PRICE_REPLACER = Pattern.compile("§6 - §a. §f([1-9]\\d*)§7" + EmeraldSymbols.E_STRING);
     private static final Pattern INGREDIENT_SPLIT_PATTERN = Pattern.compile("§f(\\d+) x (.+)");
     private static final Pattern WAR_CHAT_MESSAGE_PATTERN = Pattern.compile("§3\\[WAR§3\\] The war for (.+) will start in \\d+ minutes.");
+    private static final Pattern ABILITY_TREE_PATTERN = Pattern.compile("(Warrior|Shaman|Mage|Assassin|Archer) Abilities");
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/events/ClientEvents.java]: Does this work with reskinned classes?

File: src/main/java/com/wynntils/modules/utilities/events/ClientEvents.java [hunk 4]
```diff
@@ -1347,4 +1350,29 @@ public void onHeldItemChange(PacketEvent<CPacketHeldItemChange> e) {
         // delay because ActionBar update may be late
         new Delay(() -> cd.setElementalSpecialString(""), 4);
     }
+
+    @SubscribeEvent
+    public void onAbilityTreeScroll(GuiOverlapEvent.ChestOverlap.HandleMouseInput e) {
+        if (!Reference.onWorld) return;
+        if (!UtilitiesConfig.INSTANCE.shouldAbilityScroll) return;
+
+        int scrollAmount = Mouse.getDWheel() / 120;
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/events/ClientEvents.java]: The 120 is a magic number to me. Can you describe what it is? (and perhaps extract it into a named const for clarity)

File: src/main/java/com/wynntils/modules/utilities/events/ClientEvents.java [hunk 5]
```diff
@@ -1347,4 +1350,29 @@ public void onHeldItemChange(PacketEvent<CPacketHeldItemChange> e) {
...
+        if (scrollAmount == 0) return;
+
+        IInventory lowerInv = e.getGui().getLowerInv();
+        if (!ABILITY_TREE_PATTERN.matcher(lowerInv.getName()).matches()) return;
+
+        boolean up = scrollAmount > 0;
+        if (UtilitiesConfig.INSTANCE.invertAbilityScroll) up = !up;
+
+        int slot = up ? 57 : 59;
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/events/ClientEvents.java]: See my comment on magic numbers. (I get that these two numbers are the up and down arrow slots in the GUI, but still, extract it and give it a name.)

File: src/main/java/com/wynntils/modules/utilities/events/ClientEvents.java (hunk 3 above)
Comment (byBackfish) [reply src/main/java/com/wynntils/modules/utilities/events/ClientEvents.java]: yes, the ability tree name always takes the non-reskinned version

File: src/main/java/com/wynntils/modules/utilities/events/ClientEvents.java (hunk 4 above)
Comment (byBackfish) [reply src/main/java/com/wynntils/modules/utilities/events/ClientEvents.java]: honestly, I don't know why exactly 120, when I first looked up how to detect the direction of the mouse scroll, someone told me to divide it by 120. But it does not really make sense, as I only check if its above 0 or not, and dividing it by 120 wont change anything with that. I'll remove that 120 then

File: src/main/java/com/wynntils/modules/utilities/events/ClientEvents.java [hunk 6]
```diff
@@ -1347,4 +1353,29 @@ public void onHeldItemChange(PacketEvent<CPacketHeldItemChange> e) {
         // delay because ActionBar update may be late
         new Delay(() -> cd.setElementalSpecialString(""), 4);
     }
+
+    @SubscribeEvent
+    public void onAbilityTreeScroll(GuiOverlapEvent.ChestOverlap.HandleMouseInput e) {
+        if (!Reference.onWorld) return;
+        if (!UtilitiesConfig.INSTANCE.shouldAbilityScroll) return;
+
+        int scrollAmount = Mouse.getDWheel();
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/events/ClientEvents.java:1363]: Doesn't the event contain relevant information about scroll?

Comment (byBackfish) [reply src/main/java/com/wynntils/modules/utilities/events/ClientEvents.java:1363]: nope, it only tells you that an action with the mouse was made, and does not tell you what action

**Review by kristofbolyai — DISMISSED** [review]

File: src/main/java/com/wynntils/modules/utilities/events/ClientEvents.java (hunk 3 above)
Comment (ryanzhoudev) [inline src/main/java/com/wynntils/modules/utilities/events/ClientEvents.java]: Change to non-capturing group? Doesn't look like the class name is used

**Review by kristofbolyai — DISMISSED** [review]

File: src/main/java/com/wynntils/modules/utilities/configs/UtilitiesConfig.java [hunk 7]
```diff
@@ -79,6 +79,12 @@ public class UtilitiesConfig extends SettingsClass {
     @Setting(displayName = "Clicking on Locked Items", description = "Should moving items to and from locked inventory slots be prevented?", order = 5)
     public boolean preventSlotClicking = false;
 
+    @Setting(displayName = "Scrolling in Ability Tree", description = "Should using the scroll wheel in the ability tree move a page down/up?", order = 6)
+    public boolean shouldAbilityScroll = false;
+
+    @Setting(displayName = "Invert Scrolling in Ability Tree", description = "Should the scroll wheel controls in the ability tree be inverted?", order = 7)
+    public boolean invertAbilityScroll = false;
```
Comment (HighCrit) [inline src/main/java/com/wynntils/modules/utilities/configs/UtilitiesConfig.java]: Please use https://github.com/Wynntils/Wynntils/blob/development/src/main/java/com/wynntils/modules/core/config/CoreDBConfig.java#L37

**Review by HighCrit — DISMISSED** [review]

File: src/main/java/com/wynntils/modules/utilities/configs/UtilitiesConfig.java (hunk 7 above)
Comment (byBackfish) [reply src/main/java/com/wynntils/modules/utilities/configs/UtilitiesConfig.java]: Should I use the same config field, or should I just change the type of the `invertAbilityScroll` from a boolean to that enum?

Comment (kristofbolyai) [reply src/main/java/com/wynntils/modules/utilities/configs/UtilitiesConfig.java]: Use the enum

**Review by ryanzhoudev — DISMISSED** [review]

File: src/main/java/com/wynntils/modules/utilities/configs/UtilitiesConfig.java (hunk 7 above)
Comment (HighCrit) [reply src/main/java/com/wynntils/modules/utilities/configs/UtilitiesConfig.java]: @kristofbolyai I meant it as in the field. Such that the scroll experience is consistent across all Wynntils features

**Review by magicus — DISMISSED** [review]

**Review by magicus — APPROVED** [review]

**Review by ryanzhoudev — APPROVED** [review]

**Pull request merged**